### PR TITLE
Copyedits to Image Filters chapter

### PR DESCRIPTION
**Pull request by amyrbrown**
Here are the copyedits for this chapter. It was interesting and a fun project to read about. 

I made a fair number of changes for clarity and readability, hopefully without sacrificing the friendly, conversational tone of the writing. @catehstn, please have a look at my changes (or the final version, if that's easier) and let me know if you have any questions or concerns. I think the best way to deal with further changes at this stage is for me to make them on my copy of the repo and update this PR.

There are also a couple of comments on the Files Changed view which need some attention.

Thank you!

### CODE REVIEW

File: image-filters/image-filters.markdown [hunk 1]
```diff
@@ -3,148 +3,148 @@ author: Cate Huston
...
-## A Story of a Brilliant Idea That Wasn’t All That Brilliant
-
-In Chinese art, there is often a series of four paintings showing the same
-place in different seasons. Color - the cool whites of winter, pale hues of
-spring, lush greens of summer, and red and yellows of fall is the
-differentiation. Sometime around 2011, I had what I thought was a brilliant
-idea. I wanted to be able to visualize a photo series, as a series of colors. I
+## A Brilliant Idea (That Wasn’t All That Brilliant)
+
+Imagine a series of four paintings showing the same
```
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: I removed the reference to "Chinese art" because I did some research and couldn't find reference to this style of painting series in Chinese artistic tradition. I thought, rather than appear insensitive, it would be prudent to make this less specific.

File: image-filters/image-filters.markdown [hunk 2]
```diff
@@ -209,15 +205,15 @@ We don’t want our app to be a tiny grey window, so the two essential methods
 that we will start by overriding (implementing in our own class, instead of
```
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: Is it necessary to define "overriding" for this audience? This is maybe a question for @MichaelDiBernardo

File: image-filters/image-filters.markdown [hunk 3]
```diff
@@ -395,26 +391,26 @@ public void keyPressed() {
...
-image state incorrectly, or miscalculations of the pixel colors after applying
-a filter. I also just (some think weirdly) enjoy writing unit tests. Whilst
+image state incorrectly, or miscalculating pixel colors after applying
+a filter. I also just enjoy (some think weirdly) writing unit tests. Whilst
 some people seem to think of testing as a thing that delays checking code in, I
-see tests as my #1 debugging tool, and an opportunity to deeply understand what
+see tests as my #1 debugging tool, and as an opportunity to deeply understand what
 is going on in my code.
 
-I adore Processing, but as covered above it’s designed as a tool for artists to
```
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: Processing was described several times, so I tried to streamline the descriptions so as to reduce repetition.

File: image-filters/image-filters.markdown [hunk 4]
```diff
@@ -573,56 +570,56 @@ of colors in a program is very important to understanding how our filters work.
...
+there) are the alpha value. The values range from 00 in base 16 (0 in base
+10) through to FF (255 in base 10). The alpha represents 
+opacity, where 0 is transparent and 100% is opaque.
 
 #### HSB or HSV colors
-This color space is not quite as well known as RGB, the first number represents
-the hue, the second the saturation (how intense the color is), and the third
+This color space is not quite as well known as RGB. The first number represents
+the hue, the second number the saturation (how intense the color is), and the third
 number the brightness. The image at the top was created by manipulating in this
```
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: It's not clear what "the image at the top" refers to.

File: image-filters/image-filters.markdown [hunk 5]
```diff
@@ -573,56 +570,56 @@ of colors in a program is very important to understanding how our filters work.
...
 Having settled on this color space, it’s simpler than it would have been using
 RGB. We need to find the hue of each pixel, and figure out which is most
-"popular". We probably don’t want to be exact, we want to group very similar
+"popular". We probably don’t want to be exact — we want to group very similar
 hues together, and we can handle this using two strategies.
 
 Firstly we will round the decimals that come back to whole numbers, as this
 makes it simple to determine which "bucket" we put each pixel in. Secondly we
-can change the range of the hues. If we think back up to the diagram above, we
+can change the range of the hues. If we think back to the cone representation above, we
```
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: I wasn't sure what diagram this was, so I changed the wording --- please suggest other wording if I have missed the point!

File: image-filters/image-filters.markdown [hunk 6]
```diff
@@ -652,9 +649,9 @@ for(int px in pixels) {
 
 <markdown>
 At the end we can print this hue to the screen, or display it next to the
-picture (\aosafigref{500l.imagefilters.hueranges}).
+picture (\aosafigref{500l.imagefilters.hueranges}). 
 
-\aosafigure[240pt]{image-filters-images/hueranges.jpg}{Dominant hue vs. # buckets used in hue range}{500l.imagefilters.hueranges}
+\aosafigure[240pt]{image-filters-images/hueranges.jpg}{Dominant hue versus number of buckets used in hue range}{500l.imagefilters.hueranges}
```
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: I don't really understand this image. What do the numbers refer to? Would it be possible to write an expanded explanation of the image?

File: image-filters/image-filters.markdown [hunk 7]
```diff
@@ -667,25 +664,26 @@ picture.
...
 </markdown>
+
 <latex>
 Once we’ve extracted the "dominant" hue, we can choose to either show or hide
 it in the image. We can show the dominant hue with varying tolerance (ranges
 around it that we will accept). Pixels that don’t fall into this range can be
-changed to grayscale, by setting the value based on the brightness.
-Alternatively, we can hide the dominant hue. 
+changed to grayscale by setting the value based on the brightness.
+Alternatively, we can hide the dominant hue by redrawing pixels with that hue in greyscale and leaving other pixels as they are.
```
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: I felt this needed expanding -- is my new text correct?

File: image-filters/image-filters.markdown (hunk 2 above)
Comment (MichaelDiBernardo) [inline image-filters/image-filters.markdown]: Good question; probably not.

File: image-filters/image-filters.markdown (hunk 4 above)
Comment (MichaelDiBernardo) [inline image-filters/image-filters.markdown]: Oops. I need to turn this into a figref. I removed many images during the first editing phase, and missed this reference when doing that.

File: image-filters/image-filters.markdown (hunk 5 above)
Comment (MichaelDiBernardo) [inline image-filters/image-filters.markdown]: Again oops.

**Comment by MichaelDiBernardo** [comment]
@amyrbrown -- once this phase with @catehstn is done, I can go back and fix the image references I commented on above.

File: image-filters/image-filters.markdown (hunk 2 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: Cool; do you want me to rewrite this and push the changes to this PR? Or will you do it when you fix the image references? Either is fine by me. @MichaelDiBernardo

Comment (MichaelDiBernardo) [inline image-filters/image-filters.markdown]: I can do it when I pull refs!

File: image-filters/image-filters.markdown (hunk 1 above)
Comment (catehstn) [inline image-filters/image-filters.markdown]: sounds good - I saw these constantly in China, but also couldn't find any kind of reference to them.

File: image-filters/image-filters.markdown [hunk 8]
```diff
@@ -3,148 +3,148 @@ author: Cate Huston
 
 _Cate Huston is a developer and entrepreneur focused on mobile. She’s lived and worked in the UK, Australia, Canada, China and the United States, as an engineer at Google, an Extreme Blue intern at IBM, and a ski instructor. Cate speaks internationally on mobile development and her writing has been published on sites as varied as Lifehacker, The Daily Beast, The Eloquent Woman and Model View Culture. She co-curates Technically Speaking, blogs at Accidentally in Code and is [\@catehstn](https://twitter.com/catehstn) on Twitter._
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: Cate left the tech industry and spent a year finding her way back whilst building her passion project [Show & Hide](http://showandhide.com/). She is Director of Mobile Engineering at [Ride](https://ride.com/), speaks internationally on mobile development and engineering culture, co-curates [Technically Speaking](http://www.techspeak.email/) and is an advisor at [Glowforge](https://glowforge.com/). Cate doesn’t exactly live in Colombia but she spends a lot of time there, and has lived and worked in the UK, Australia, Canada, China the United States, previously as an engineer at Google, an Extreme Blue intern at IBM, and a ski instructor. Cate blogs at [Accidentally in Code](http://accidentallyincode.com/) and is [@catehstn](https://twitter.com/catehstn) on Twitter.

File: image-filters/image-filters.markdown [hunk 9]
```diff
@@ -209,15 +205,15 @@ We don’t want our app to be a tiny grey window, so the two essential methods
...
-by calling `redraw()` (as covered in the Processing Documentation, `draw()`
-should not be called explicitly). 
+[`draw()`](http://processing.org/reference/draw_.html). The `setup()` method is only
+called when the app starts, and is where we do things like set the size of 
+the app window.  The `draw()` method is called for every animation, or after some action can be triggered
+by calling `redraw()`. (As covered in the Processing Documentation, `draw()`
+should not be called explicitly.) 
 
 Processing is designed to work nicely to create animated sketches, but in this
 case we don’t want animation[^noanim], we want to respond to key presses. To prevent
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: what does [^noanim] do?

File: image-filters/image-filters.markdown [hunk 10]
```diff
@@ -573,56 +570,56 @@ of colors in a program is very important to understanding how our filters work.
...
 
-You may not have known it at the time, but we were unwittingly using a concept
-in the previous section called a "color space", which is way of representing
-color digitally. Kids mixing paints learn that all colors can be made from
-other colors, but things work slightly differently in digital (less risk of
+We were using a concept
+in the previous section called "color space", which is way of representing
+color digitally. Kids mixing paints learn that colors can be made from
+other colors; things work slightly differently in digital (less risk of
 being covered in paint!) but similar. Processing makes it really easy to work
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: should this be "similarly".

File: image-filters/image-filters.markdown (hunk 1 above)
Comment (catehstn) [inline image-filters/image-filters.markdown]: Maybe we could reference it as something I saw rather than like, a general concept?

File: image-filters/image-filters.markdown [hunk 11]
```diff
@@ -573,56 +570,56 @@ of colors in a program is very important to understanding how our filters work.
...
-red the second two blue, the third two green, and the final two (if they are
-there) are the alpha value. These values range from 00 in base 16, 0 in base
-10, through to FF, the equivalent of 255 in base 10. The alpha represents the
-opacity, where 0 is transparent, and 100% opaque.
+The color space that most programmers are familiar with is RGBA: red, green,
+blue and alpha. In hexadecimal (base 16), the first two digits are the amount of
+red, the second two blue, the third two green, and the final two (if they are
+there) are the alpha value. The values range from 00 in base 16 (0 in base
+10) through to FF (255 in base 10). The alpha represents 
+opacity, where 0 is transparent and 100% is opaque.
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: this was already explained above. should we note that?

File: image-filters/image-filters.markdown (hunk 6 above)
Comment (catehstn) [inline image-filters/image-filters.markdown]: Sure, it's like:
`colorMode(HSB, numberOfBuckets);`
is that clearer?

File: image-filters/image-filters.markdown [hunk 12]
```diff
@@ -667,25 +664,26 @@ picture.
 Once we’ve extracted the "dominant" hue, we can choose to either show or hide
 it in the image. We can show the dominant hue with varying tolerance (ranges
 around it that we will accept). Pixels that don’t fall into this range can be
-changed to grayscale, by setting the value based on the brightness.
+changed to grayscale by setting the value based on the brightness.
 \aosafigref{500l.imagefilters.showdominant} shows the dominant hue determined
 using a hue range of 240, and with varying tolerance. The tolerance is the
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: i think part of the problem is that I'm using different words for "hue range" / "number of buckets" - should probably clarify that number of buckets == hue range, which is the colorMode argument.

File: image-filters/image-filters.markdown (hunk 7 above)
Comment (catehstn) [inline image-filters/image-filters.markdown]: I wouldn't call it redrawing but setting the color. Otherwise, yes!

File: image-filters/image-filters.markdown [hunk 13]
```diff
@@ -773,36 +771,37 @@ matrices of specific values that sharpen images.
...
+processing and filtering takes place, and `PixelColorHelper.java` 
 abstracts out final `PApplet` methods for pixel colors for testability.
 
 \aosafigure[240pt]{image-filters-images/architecture.jpg}{Architecture diagram}{500l.imagefilters.architecture}
 
 ### Wrapper Classes and Tests
 Briefly mentioned above, there are two wrapper classes (`IFAImage` and
-`PixelColorHelper`) that wrap library methods for testability. This is because
+`PixelColorHelper`) that wrap library methods for testability. This is because,
 in Java, final methods are methods that cannot be overridden or hidden by
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: I would rephrase this to say, "in Java, the keyword final indicates a method that cannot be overridden..."

File: image-filters/image-filters.markdown [hunk 14]
```diff
@@ -1064,7 +1063,7 @@ public class ColorHelper {
 ```
 
 Clearly we can’t test this with whole images. Instead we can mock the images
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: I think this should change - maybe to, "We don't want to test this with whole images, because we want images that we know the properties of and reason about. We approximate this by mocking the images and making them return an array of pixels..."

Comment (catehstn) [inline image-filters/image-filters.markdown]: Actually when I ported this to Obj-C, I did make something that would make actual images with certain properties and allow me to compare them, so they didn't have to be mocked. https://github.com/catehstn/image-creator

File: image-filters/image-filters.markdown [hunk 15]
```diff
@@ -1552,16 +1551,18 @@ Making things look just so. Making them fail over. Maintaining that 99.9%
 uptime. We spend more time hunting down corner cases than refining algorithms.
```
Comment (catehstn) [inline image-filters/image-filters.markdown]: line above - "fail over" - I think someone else changed this, because I had to look up what it means. It's actually "failover" and maybe we should link to a definition?

File: image-filters/image-filters.markdown (hunk 1 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: Sure, how about "When I was traveling in China I often saw series of four paintings showing..." ?

File: image-filters/image-filters.markdown (hunk 10 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: I left it the way it is because it's conversational. If we wanted to be very proper I would remove "but similar" altogether, because it doesn't add much to "slightly differently". I think it's fine the way it is.

File: image-filters/image-filters.markdown (hunk 11 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: Sure -- I'll make the change.

File: image-filters/image-filters.markdown (hunk 12 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: Okay -- I'll change the text to clarify this (I hope) and push the changes to this PR.

File: image-filters/image-filters.markdown (hunk 13 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: Okay.

File: image-filters/image-filters.markdown (hunk 14 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: In the interest of brevity and simplicity, I think we should just use the text you suggested above rather than introduce your new tool. @MichaelDiBernardo?

File: image-filters/image-filters.markdown (hunk 15 above)
Comment (amyrbrown) [inline image-filters/image-filters.markdown]: If it's not your text I would remove it --- it doesn't really make sense in the context of this chapter anyway; it's not like you're going to failover test an image filter app...  Can you think of a more appropriate thing to add to the list of "stuff we do in productionisation"? Maybe "Testing edge cases."?

File: image-filters/image-filters.markdown (hunk 1 above)
Comment (catehstn) [inline image-filters/image-filters.markdown]: yes! love it!

File: image-filters/image-filters.markdown (hunk 10 above)
Comment (catehstn) [inline image-filters/image-filters.markdown]: OK

File: image-filters/image-filters.markdown (hunk 15 above)
Comment (catehstn) [inline image-filters/image-filters.markdown]: "testing edge cases" is covered in the next sentence. i would just delete it.

**Comment by MichaelDiBernardo** [comment]
To save a bunch of githubbery, I am going to merge this now and handle the outstanding comments separately. You both have done a great job of taking care of all the things that require actual thinking, which makes my job a lot easier -- thank you!

File: image-filters/image-filters.markdown (hunk 4 above)
Comment (MichaelDiBernardo) [inline image-filters/image-filters.markdown]: Turns out this was referring to an image I removed. I'll remove this reference as well.

**Comment by MichaelDiBernardo** [comment]
OK, I made the final round of changes in https://github.com/aosabook/500lines/commit/19ca503aaeac88872f7427876e8c3ae3fc3164ac. Thanks again for all your hard work here @catehstn and @amyrbrown!
